File: Blind75/124. Binary Tree Maximum Path Sum/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = INT_MIN;
    unordered_map<TreeNode*, int> memo;
    int maxPathSum(TreeNode* root) {
        helper(root);
        return res;
    }

    int helper(TreeNode* node){
        //base case
        if(node==NULL) return 0;
        if(memo.count(node)) return memo[node];
        int leftTree = max(helper(node->left),0);
        int rightTree = max(helper(node->right),0);

        int curr_max = node->val+leftTree+rightTree;
        res = max(res, curr_max);
        return memo[node] = node->val + max(leftTree, rightTree);
    }
};

File: 124. Binary Tree Maximum Path Sum/124. Binary Tree Maximum Path Sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int res = Int32.MinValue;
    Dictionary<TreeNode, int> dect = new Dictionary<TreeNode, int>();
    public int MaxPathSum(TreeNode root) {
        dp(root);
        return res;
    }

    int dp(TreeNode node){
        //base case
        if(node==null) return 0;
        if(dect.ContainsKey(node)) return dect[node];

        var left = Math.Max(dp(node.left), 0);
        var right = Math.Max(dp(node.right), 0);

        int curr_max = node.val + left + right;
        res = Math.Max(curr_max, res);

        dect.Add(node, node.val + Math.Max(left,right));
        return dect[node];

    }
}
